File: DailyChallenges/Scala/LeetCode_2215.scala
object Solution {
    def findDifference(nums1: Array[Int], nums2: Array[Int]): List[List[Int]] = {
        val ans = Array.fill[List[Int]](2)(Nil)
        def helper(l1: List[Int], l2: List[Int]): Unit = {
            if (l1 == Nil && l2 == Nil) return
            if (l1 == Nil) ans(1) ++= l2
            else if (l2 == Nil) ans(0) ++= l1
            else (l1.head, l2.head) match {
                case (h1, h2) if h1 == h2 => helper(l1.tail, l2.tail)
                case (h1, h2) if h1 < h2 => ans(0) = h1 :: ans(0); helper(l1.tail, l2)
                case _ => ans(1) = l2.head :: ans(1); helper(l1, l2.tail)
            }
        }
        helper(nums1.sorted.distinct.toList, nums2.sorted.distinct.toList)
        ans.toList
    }
}

File: DailyChallenges/CPP/LeetCode_2215.cpp
class Solution {
public:
    vector<vector<int>> findDifference(vector<int>& nums1, vector<int>& nums2) {
        vector<vector<int>> ans(2);
        set<int> s1(nums1.begin(), nums1.end()), s2(nums2.begin(), nums2.end());
        auto itr1 = s1.begin(), itr2 = s2.begin();
        while (itr1 != s1.end() || itr2 != s2.end()){
            if (itr1 == s1.end()){
                while (itr2 != s2.end()){
                    ans[1]. push_back(*itr2++);
                }
            } else if (itr2 == s2.end()){
                while (itr1 != s1.end()){
                    ans[0]. push_back(*itr1++);
                }
            } else if (*itr1 == *itr2){
                ++itr1; ++itr2;
            } else if (*itr1 < *itr2){
                ans[0].push_back(*itr1++);
            }else {
                ans[1].push_back(*itr2++);
            }
        }
        return ans;
    }
};
